libvast/src/address.cpp

```diff
@@ -16,13 +16,104 @@
 
 #include <arpa/inet.h>
 #include <netinet/in.h>
+#include <openssl/crypto.h>
+#include <openssl/err.h>
+#include <openssl/evp.h>
+#include <openssl/ossl_typ.h>
 #include <sys/socket.h>
 
 #include <cstdlib>
 #include <utility>
 
 namespace vast {
 
+namespace {
+
+inline uint32_t bitmask32(size_t bottom_bits) {
+  return bottom_bits >= 32 ? 0xffffffff : ((uint32_t{1} << bottom_bits) - 1);
+}
+
+class address_encryptor {
+public:
+  address_encryptor(const std::array<address::byte_type, 32>& key) {
+    ctx_ = EVP_CIPHER_CTX_new();
+    EVP_CIPHER_CTX_init(ctx_);
+    OpenSSL_add_all_ciphers();
+    cipher_ = EVP_get_cipherbyname("aes-128-ecb");
+    block_size_ = EVP_CIPHER_block_size(cipher_);
+    pad_ = std::vector<address::byte_type>(block_size_);
+    auto pad_out_len = 0;
+    EVP_CipherInit_ex(ctx_, cipher_, nullptr, key.data(), nullptr, 1);
+    // use second 16-byte half of key for padding
+    EVP_CipherUpdate(ctx_, pad_.data(), &pad_out_len, key.data() + block_size_,
+                     block_size_);
+  }
+
+  address_encryptor(const address_encryptor&) = delete;
+
+  ~address_encryptor() {
+    EVP_CIPHER_CTX_cleanup(ctx_);
+    EVP_CIPHER_CTX_free(ctx_);
+  }
+
+  auto operator()(address::byte_array bytes, size_t byte_offset) {
+    auto bytes_to_encrypt = std::span{bytes.begin() + byte_offset, bytes.end()};
+    auto one_time_pad = generate_one_time_pad(bytes_to_encrypt);
+    for (auto i = size_t{0}; i < bytes_to_encrypt.size(); ++i) {
+      bytes[i + byte_offset] = bytes_to_encrypt[i] ^ one_time_pad[i];
+    }
+    return bytes;
+  }
+
+private:
+  static constexpr inline auto msb_of_byte_mask = 0b10000000;
+  EVP_CIPHER_CTX* ctx_;
+  const EVP_CIPHER* cipher_;
+  int block_size_;
+  std::vector<address::byte_type> pad_;
```

Comment:
```suggestion
  EVP_CIPHER_CTX* ctx_ = {};
  const EVP_CIPHER* cipher_ {};
  int block_size_ = {};
  std::vector<address::byte_type> pad_ = {};
```

Even though it's not strictly needed for the vector, it is very much a good habit to get into for all member variables. The first three are trivially constructible and may as such contain garbage values if you don't explicitly initialize them.